src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala

```diff
@@ -81,38 +121,65 @@ class LibraryService (protected val argUserInfo: UserInfo, val rawlsDAO: RawlsDA
+                // between the time we retrieved them and here, where we update them.
+                val allOperations = generateAttributeOperations(workspaceResponse.workspace.attributes, userAttrs,
+                  k => k.namespace == AttributeName.libraryNamespace && k.name != LibraryService.publishedFlag.name)
+                RequestComplete(patchWorkspace(ns, name, allOperations, isPublished(workspaceResponse)))
+              } else {
+                RequestCompleteWithErrorReport(Forbidden, s"must have at least write permissions")
               }
             }
           }
         }
     }
   }
 
+  /*
+   * Update workspace in rawls. If the dataset is currently published, republish with changes
+   */
+  def patchWorkspace(ns: String, name: String, allOperations: Seq[AttributeUpdateOperation], isPublished: Boolean) : Future[Workspace] = {
+    rawlsDAO.patchWorkspaceAttributes(ns, name, allOperations) map { newws =>
+      if (isPublished) {
+        // if already published, republish
+        // we do not need to delete before republish
+        publishDocument(newws)
+      }
+      newws
+    }
+  }
+
+  def canPublish(workspaceResponse: WorkspaceResponse): Future[Boolean] = {
+    if (isPublished(workspaceResponse)) {
+      // workspace is already published
+      // republish just needs write
+      hasAccessOrCurator(workspaceResponse, WorkspaceAccessLevels.Write)
+    } else {
+      hasAccessOrCurator(workspaceResponse, WorkspaceAccessLevels.Owner)
+    }
+  }
+
   def setWorkspaceIsPublished(ns: String, name: String, value: Boolean): Future[PerRequestMessage] = {
     rawlsDAO.getWorkspace(ns, name) flatMap { workspaceResponse =>
-      // verify owner on workspace
-      if (workspaceResponse.accessLevel < WorkspaceAccessLevels.Owner) {
-        Future(RequestCompleteWithErrorReport(Forbidden, "must be an owner"))
-      } else {
-        val operations = updatePublishAttribute(value)
-        rawlsDAO.patchWorkspaceAttributes(ns, name, operations) map { ws =>
-          if (value)
-            publishDocument(ws)
+      canPublish(workspaceResponse) flatMap { canPublish =>
+        if (canPublish) {
+          // has access
+          val operations = updatePublishAttribute(value)
+          rawlsDAO.patchWorkspaceAttributes(ns, name, operations) map { ws =>
+            if (value)
+              publishDocument(ws)
+            else
+              removeDocument(ws)
+            RequestComplete(ws)
+          }
+        } else {
+          // verify owner on workspace
+          if (workspaceResponse.accessLevel < WorkspaceAccessLevels.Write)
+            Future(RequestCompleteWithErrorReport(Forbidden, "must be a curator or have at least write privileges"))
           else
-            removeDocument(ws)
-          RequestComplete(ws)
+            Future(RequestCompleteWithErrorReport(Forbidden, "must be an owner or a curator"))
```

Comment:
refactored this code. the removeDocument is for unpublishing not addressing 1683, but admit code was confusing. hopefully my changes are clearer, even if duplicating some code. let me know if you have a better way :)